EIPS/eip-draft.md

```diff
@@ -0,0 +1,226 @@
+
+### Implementation Pseudocode
+
+```python
+# Before transaction execution:
+W = set()
+for tuple in tx.access_list:
+    warm(tuple.address)  # per EIP-2930/EIP-2929 rules
+    if tuple.checkCodeHash == true:
+        acc = load_account(tuple.address)
+        if acc exists and acc.code is not empty:
+            W.add(acc.codeHash)
+
+# On successful CREATE/CREATE2:
+H = keccak256(B)
+if H in W:
+    # Duplicate: no deposit gas
+    link_codehash(new_account, H)
+else:
+    # New bytecode: charge and persist
+    charge(GAS_CODE_DEPOSIT * len(B))
+    persist_code(H, B)
+    link_codehash(new_account, H)
+```
+
+## Rationale
+
+### Why Access-List Based Deduplication?
+
+The access-list approach provides several critical properties:
+
+1. Deterministic behavior: 
+The result depends only on the transaction's access list and current state, not on local database contents. All nodes compute the same gas cost.
+
+2. No reverse index requirement: 
+Unlike other approaches, this doesn't require maintaining a `codeHash → [accounts]` reverse index, which would add significant complexity and storage overhead.
+
+3. Leverages existing infrastructure: 
+Builds on EIP-2930 access lists and EIP-2929 access costs, requiring minimal protocol changes.
+
+4. Explicit opt-in: 
+Transactions must explicitly indicate which addresses to check. This prevents unexpected behavior and gives users/wallets control over gas optimization.
+
+5. Forward compatibility: 
+Pre-fork nodes ignore `checkCodeHash` and never grant discounts. Post-fork, all nodes enforce identical behavior. Wallets can optionally add the field to optimize gas, but its absence doesn't invalidate transactions.
+
+6. Avoiding have a code-root for state:
+At this point, clients handle code storage on their own ways. They don't have any consensus on the deployed existing codes (besides that all of the ones referenced in account's codehash fields exist).
+Changing this seems a lot more complex and unnecessary.
+
+### Same-Block Deployments
+
+Sequential transaction execution ensures that a deployment storing new code makes it visible to later transactions in the same block:
+
+1. Transaction `T_A` deploys bytecode `B` at address `X`
+   - Pays full `GAS_CODE_DEPOSIT * L` (no prior contract has this bytecode)
+   - Code is stored under hash `H = keccak256(B)`
+
+2. Later transaction `T_B` in the same block deploys the same bytecode `B`:
+   - `T_B` includes address `X` in its access list with `checkCodeHash: true`
+   - When `T_B` executes, `W` is built from the current state (including `T_A`'s changes)
+   - Since `X` now exists, `W` contains `H`
+   - `T_B`'s deployment gets the discount
+
+> While this only tries to formalize the behaviour, it's important to remark that this kind of behaviour is complex. As it requires control over tx ordering in order to abuse. And Builders can't modify the Acess List as it is already signed with the Tx. Nevertheless, this could happen, thus is formalized here. 
```

Comment:
I'm being pedantic here, but either you "formalise the behaviour" or you "formalize the behavior"